File: funNLearn/src/main/java/dsAlgo/sort/QuickSort.java
package dsAlgo.sort;

public class QuickSort
{
   static void quick_srt(int arr[], int left, int right) {
      int index = partition(arr, left, right);
     
      if (left < index - 1)
         quick_srt(arr, left, index - 1);
      if (index < right)
         quick_srt(arr, index, right);
   }
  
   static int partition(int arr[], int left, int right)
   {
         int i = left, j = right;
         int tmp;
         int pivot = arr[(left + right) / 2];
       
         while (i <= j)
         {
               while (arr[i] < pivot)
                     i++;
               while (arr[j] > pivot)
                     j--;
               if (i <= j) {
                     tmp = arr[i];
                     arr[i] = arr[j];
                     arr[j] = tmp;
                     i++;
                     j--;
               }
         };
         return i;
   }
  
   public static void print(int[] myArr)
   {
      for(int i=0; i<myArr.length; i++)
         System.out.print(myArr[i]+ ", ");
      System.out.println();
   } 
  
   public static void main(String a[])
   {
      int array[] = {2, 8, 7, 6, 4, 1, 8, 5, 3, 2, 10};
     
      System.out.println("Values Before the sort:");
      print(array);
     
      quick_srt(array, 0, array.length - 1);
     
      System.out.print("\nValues after the sort:\n");
      print(array);
   }
}

File: funNLearn/src/main/java/dsAlgo/sort/QuickSort.py
"""
    Tag: sort
"""
from typing import List


class Solution:
    def partition(self, A: List[int], low: int, high: int) -> int:
        i = low - 1
        pivot = A[high]

        for j in range(low, high):
            if A[j] <= pivot:
                i = i + 1
                A[i], A[j] = A[j], A[i]

        A[i + 1], A[high] = A[high], A[i + 1]
        return i + 1

    def quickSort(self, A: List[int], low: int, high: int):
        """
            -  Pick a pivot, usually the last element in the array. 
            -  start from first element and move the small elements to left and 
            larger elements to right. 
            -  Find the right location for the pivot so that everything on it's 
            left is smaller and everything on it's right is bigger
            -  break the array from pivot. repeat the process for left and right halves. 
        """
        if low < high:
            pi = self.partition(A, low, high)
            self.quickSort(A, low, pi - 1)
            self.quickSort(A, pi + 1, high)

    def QS(self, A: List[int]) -> List[int]:
        self.quickSort(A, 0, len(A) - 1)
        return A


assert Solution().QS([3, 2, 9, 3, 5, 7, 1, 8]) == [1, 2, 3, 3, 5, 7, 8, 9]
print("Tests Passed!")
